src/api/app/controllers/webui/packages/files_controller.rb

```diff
@@ -0,0 +1,75 @@
+module Webui
+  module Packages
+    class FilesController < Packages::MainController
+      before_action :set_project
+      before_action :set_package
+      after_action :verify_authorized
+
+      def new
+        authorize @package, :update?
+      end
+
+      def create
+        authorize @package, :update?
+
+        file = params[:file]
+        file_url = params[:file_url]
+        filename = params[:filename]
+
+        errors = []
+
+        begin
+          if file.present?
+            # We are getting an uploaded file
+            filename = file.original_filename if filename.blank?
+            @package.save_file(file: file, filename: filename, comment: params[:comment])
+          elsif file_url.present?
+            # we have a remote file URI, so we have to download and save it
+            services = @package.services
+
+            # detects automatically git://, src.rpm formats
+            services.addDownloadURL(file_url, filename)
+
+            unless services.save
+              errors << "Failed to add file from URL '#{file_url}'"
+            end
+          elsif filename.present? # No file is provided so we just create an empty new file (touch)
+            @package.save_file(filename: filename)
+          else
+            errors << 'No file or URI given'
+          end
+        rescue APIError => e
+          errors << e.message
+        rescue Backend::Error => e
+          errors << Xmlhash::XMLHash.new(error: e.summary)[:error]
+        rescue StandardError => e
+          errors << e.message
+        end
+
+        if errors.empty?
+          message = "The file '#{filename}' has been successfully saved."
+          # We have to check if it's an AJAX request or not
```

Comment:
This comment is quite useless. I would remove it. Same for the one on line 60.

Reply:
This code was only moved to this controller. I plan to refactor the whole `create` action in another PR.